### ISSUE TITLE
Not recognizing final keyword in class declaration

### ISSUE THREAD

**Issue by aldenml** [swig/swig#672]
Hi, now that c++ 11 is getting more adoption, I have found the following problem:
C++ header file (demo_final.hpp)

```
struct demo_class_test final
{
}
```

Swig file (demo_final.i)

```
%module demoFinal
%include "demo_final.hpp"
```

Command:

```
$ swig -c++ -java demo_final.i
```

Result:

```
demo_final.hpp:3: Warning 314: 'final' is a java keyword, renaming to '_final'
```

Any comment about supporting (or how to support) this keyword at class declaration? Thanks

**Comment by ojwb** [comment]
It looks like the parser in `Source/CParse/parser.y` needs extending to allow `final` here (it's only currently supported for methods AFAICS).

If you want `final` to actually change anything in the generated code (e.g. disabling directors for the class) then it'll need adding to the node for the class in the parse tree, and then handling at code generation time.

**Comment by n8vm** [comment]
Running into this issue 2 years later. What's the status on this?

**Comment by ojwb** [comment]
Nobody's worked on it yet it seems.  Feel free to work up a patch if it's something that particularly matters to you.

A simpler workaround is just to run swig with `-Dfinal=` to make final disappear entirely.  That shouldn't make any difference to other cases where `final` is supported by SWIG as it isn't currently added to the parse tree even in those cases anyway.

**Comment by n8vm** [comment]
Shouldn't "final" just be ignored when parsing? Perhaps with a warning? What needs changed in parser.y? 

Here? Possibly after "inherit" ?
https://github.com/swig/swig/blob/6f2399e86da13a9feb436e3977e15d2b9738294e/Source/CParse/parser.y#L3643

**Comment by ojwb** [comment]
Quite possibly, unless that introduces grammar conflicts.  It would really need test coverage adding too.

**Comment by romintomasetti** [comment]
I'm not an expert in writing patches in the `parser.y`, but as @ojwb mentioned, it needs a test suite.

It can be easily done by updating `Examples/python/class/example.h` (just appending a new class to the file):

```c++
class Triangle final : public Shape {
private:
    double _angle_1,_angle_2;
    double _side_1;
public:
    Triangle(const double& angle_1, const double& angle_2, const double& side_1)
    : _angle_1(angle_1), _angle_2(angle_2), _side_1(side_1) {}
    double area() override;
    double perimeter() override;
};
```

As of SWIG ac8de714af1d7fbd755597b0f51792212f552ef7 this fails to be parsed successfully with the following error message:

```bash
example.h:36: Error: Syntax error in input(1).
```

I think that a good fix is simply to ignore the `final` keyword.

**Comment by ptizoom** [comment]
> I think that a good fix is simply to ignore the `final` keyword. 

nop, I think, because of "director" and the class will missmatch at compile ?
is there some work/branch  in progress for this matter?

**Comment by ptizoom** [comment]
> Maybe something like 
cpp_class_decl : storage_class cpptype idcolon inherit _specifier_ LBRACE

**virt_specifier_seq** needs to be used before `idcolon` I guess?

also seen like in `exception_specification`,
or more interrestingly... like in `c_decl `
```
           /* Alternate function syntax introduced in C++11:
              auto funcName(int x, int y) -> int; */
           | storage_class AUTO declarator cpp_const ARROW cpp_alternate_rettype **virt_specifier_seq_opt** initializer c_decl_tail {

```
tbc...

**Comment by ojwb** [comment]
> > I think that a good fix is simply to ignore the `final` keyword.
> 
> nop, I think, because of "director" and the class will missmatch at compile ?

Directors allow subclassing a C++ class in the scripting language such that virtual methods get magically routed as you'd naturally expect - conceptually `final` on a class means "this class cannot be subclassed", so directors don't make sense for such a class (and the implementation of directors relies on SWIG being able to create a C++ subclass of the class, which isn't possible for a class marked `final`.)

Ideally SWIG would probably automatically translate `final` on a class to `feature:nodirector` (or effectively so), but if SWIG just ignored `final` on a class then I think everything would just work if directors weren't enabled at all, and the user could explicitly mark the class with `%nodirector` to have it work when directors were enabled.

> is there some work/branch in progress for this matter?

The only work I'm aware of is the comments in this ticket.

Generally changes happen in SWIG because somebody cares enough to come up with a clean patch with test coverage which implements them.  If this is something you care about, that somebody could be you!

**Comment by ojwb** [comment]
> It can be easily done by updating `Examples/python/class/example.h` (just appending a new class to the file):

We want this in the testsuite proper really - `Examples/test-suite/`.  And it needs to be in a C++11-specfic file (one whose name starts `cpp11_`).  If we add it where you suggest that file would no longer compile as C++98.  But the testsuite part is pretty easy, as we're only checking SWIG parses it successfully while there's no actual run-time effect as such.

> virt_specifier_seq needs to be used before inherit sounds better , right?

It'd need to be `virt_specifier_seq_opt` since the `final` needs to be optional, but `virt_specifier_seq` allows `OVERRIDE` so isn't right here anyway, see: https://alx71hub.github.io/hcb/#class-virt-specifier-seq

I had a quick go at adding this, but I get an extra shift/reduce conflict which we're keen to avoid.  Going to give up for now, but here's the patch I tried in case there are any bison wizards reading:

```
diff --git a/Source/CParse/parser.y b/Source/CParse/parser.y
index c487f3b9..d2a4d7a7 100644
--- a/Source/CParse/parser.y
+++ b/Source/CParse/parser.y
@@ -1700,6 +1700,7 @@ static String *add_qualifier_to_declarator(SwigType *type, SwigType *qualifier)
 %type <node>     lambda_introducer lambda_body;
 %type <pl>       lambda_tail;
 %type <str>      virt_specifier_seq virt_specifier_seq_opt;
+%type <str>      class_virt_specifier_seq_opt;
 
 %%
 
@@ -3672,7 +3673,7 @@ cpp_declaration : cpp_class_decl {  $$ = $1; }
 
 
 /* A simple class/struct/union definition */
-cpp_class_decl  : storage_class cpptype idcolon inherit LBRACE {
+cpp_class_decl: storage_class cpptype idcolon class_virt_specifier_seq_opt inherit LBRACE {
                    String *prefix;
                    List *bases = 0;
 		   Node *scope = 0;
@@ -3680,10 +3681,10 @@ cpp_class_decl  : storage_class cpptype idcolon inherit LBRACE {
 		   $<node>$ = new_node("class");
 		   Setline($<node>$,cparse_start_line);
 		   Setattr($<node>$,"kind",$2);
-		   if ($4) {
-		     Setattr($<node>$,"baselist", Getattr($4,"public"));
-		     Setattr($<node>$,"protectedbaselist", Getattr($4,"protected"));
-		     Setattr($<node>$,"privatebaselist", Getattr($4,"private"));
+		   if ($5) {
+		     Setattr($<node>$,"baselist", Getattr($5,"public"));
+		     Setattr($<node>$,"protectedbaselist", Getattr($5,"protected"));
+		     Setattr($<node>$,"privatebaselist", Getattr($5,"private"));
 		   }
 		   Setattr($<node>$,"allows_typedef","1");
 
@@ -3711,8 +3712,8 @@ cpp_class_decl  : storage_class cpptype idcolon inherit LBRACE {
 		   Setattr($<node>$, "Classprefix", $3);
 		   Classprefix = NewString($3);
 		   /* Deal with inheritance  */
-		   if ($4)
-		     bases = Swig_make_inherit_list($3,Getattr($4,"public"),Namespaceprefix);
+		   if ($5)
+		     bases = Swig_make_inherit_list($3,Getattr($5,"public"),Namespaceprefix);
 		   prefix = SwigType_istemplate_templateprefix($3);
 		   if (prefix) {
 		     String *fbase, *tbase;
@@ -3792,7 +3793,7 @@ cpp_class_decl  : storage_class cpptype idcolon inherit LBRACE {
 		   Delattr($$, "prev_symtab");
 		   
 		   /* Check for pure-abstract class */
-		   Setattr($$,"abstracts", pure_abstracts($7));
+		   Setattr($$,"abstracts", pure_abstracts($8));
 		   
 		   /* This bit of code merges in a previously defined %extend directive (if any) */
 		   {
@@ -3808,12 +3809,12 @@ cpp_class_decl  : storage_class cpptype idcolon inherit LBRACE {
 		   scpname = Swig_symbol_qualifiedscopename(0);
 		   Setattr(classes, scpname, $$);
 
-		   appendChild($$, $7);
+		   appendChild($$, $8);
 		   
 		   if (am) 
 		     Swig_extend_append_previous($$, am);
 
-		   p = $9;
+		   p = $10;
 		   if (p && !nscope_inner) {
 		     if (!cparse_cplusplus && currentOuterClass)
 		       appendChild(currentOuterClass, p);
@@ -3837,8 +3838,8 @@ cpp_class_decl  : storage_class cpptype idcolon inherit LBRACE {
 		     }
 		     p = nextSibling(p);
 		   }
-		   if ($9 && Cmp($1,"typedef") == 0)
-		     add_typedef_name($$, $9, $3, cscope, scpname);
+		   if ($10 && Cmp($1,"typedef") == 0)
+		     add_typedef_name($$, $10, $3, cscope, scpname);
 		   Delete(scpname);
 
 		   if (cplus_mode != CPLUS_PUBLIC) {
@@ -3860,12 +3861,12 @@ cpp_class_decl  : storage_class cpptype idcolon inherit LBRACE {
 		   if (cplus_mode == CPLUS_PRIVATE) {
 		     $$ = 0; /* skip private nested classes */
 		   } else if (cparse_cplusplus && currentOuterClass && ignore_nested_classes && !GetFlag($$, "feature:flatnested")) {
-		     $$ = nested_forward_declaration($1, $2, $3, Copy($3), $9);
+		     $$ = nested_forward_declaration($1, $2, $3, Copy($3), $10);
 		   } else if (nscope_inner) {
 		     /* this is tricky */
 		     /* we add the declaration in the original namespace */
 		     if (Strcmp(nodeType(nscope_inner), "class") == 0 && cparse_cplusplus && ignore_nested_classes && !GetFlag($$, "feature:flatnested"))
-		       $$ = nested_forward_declaration($1, $2, $3, Copy($3), $9);
+		       $$ = nested_forward_declaration($1, $2, $3, Copy($3), $10);
 		     appendChild(nscope_inner, $$);
 		     Swig_symbol_setscope(Getattr(nscope_inner, "symtab"));
 		     Delete(Namespaceprefix);
@@ -3877,14 +3878,14 @@ cpp_class_decl  : storage_class cpptype idcolon inherit LBRACE {
 		     Swig_symbol_setscope(cscope);
 		     Delete(Namespaceprefix);
 		     Namespaceprefix = Swig_symbol_qualifiedscopename(0);
-		     add_symbols($9);
+		     add_symbols($10);
 		     if (nscope) {
 		       $$ = nscope; /* here we return recreated namespace tower instead of the class itself */
-		       if ($9) {
-			 appendSibling($$, $9);
+		       if ($10) {
+			 appendSibling($$, $10);
 		       }
 		     } else if (!SwigType_istemplate(ty) && template_parameters == 0) { /* for template we need the class itself */
-		       $$ = $9;
+		       $$ = $10;
 		     }
 		   } else {
 		     Delete(yyrename);
@@ -3895,7 +3896,7 @@ cpp_class_decl  : storage_class cpptype idcolon inherit LBRACE {
 			 outer = Getattr(outer, "nested:outer");
 		       appendSibling(outer, $$);
 		       Swig_symbol_setscope(cscope); /* declaration goes in the parent scope */
-		       add_symbols($9);
+		       add_symbols($10);
 		       set_scope_to_global();
 		       Delete(Namespaceprefix);
 		       Namespaceprefix = Swig_symbol_qualifiedscopename(0);
@@ -3908,7 +3909,7 @@ cpp_class_decl  : storage_class cpptype idcolon inherit LBRACE {
 		     } else {
 		       yyrename = Copy(Getattr($<node>$, "class_rename"));
 		       add_symbols($$);
-		       add_symbols($9);
+		       add_symbols($10);
 		       Delattr($$, "class_rename");
 		     }
 		   }
@@ -6987,6 +6988,14 @@ virt_specifier_seq_opt : virt_specifier_seq {
                }
                ;
 
+class_virt_specifier_seq_opt : FINAL {
+                   $$ = NewString("1")
+               }
+               | empty {
+                   $$ = 0;
+               }
+               ;
+
 exception_specification : THROW LPAREN parms RPAREN {
                     $$.throws = $3;
                     $$.throwf = NewString("1");
```

I also did a quick test with duplicating `cpp_decl` and inserting `FINAL` into one of them, but that also gives a shift/reduce conflict.

**Comment by ojwb** [comment]
It just clicked with me what this conflict is from.

If I (as a test - not proposing this as the solution) remove `FINAL` from `identifier` then I get rid of the conflict from the patch above (which we then see is missing a `;` after `$$ = NewString("1")`).

Take the example in the original report above:

```
struct demo_class_test final
{
};
```

This doesn't give a parse error because it parses like this:

```
class A a{};
```

i.e. it's defining a variable called `final` of type `struct demon_class_test` with an empty initialiser.  They made `final` a special type of keyword which is still a valid identifier in contexts where it's not allowed as a keyword.

I'm not quite sure how to avoid this conflict.  Maybe we allow another shift/reduce conflict but ensure it's resolved the correct way.  Otherwise it seems we need to avoid `final` being a valid variable name here under some conditions that seems complicated to implement.

@wsfulton Thoughts?

**Comment by wsfulton** [comment]
Years ago I also came up with I think was the same solution but it introduced one shift/reduce for adding in class final. If I recall correctly there are other areas where final support is missing and yet more shift/reduce conflicts appeared. Your analysis is interesting because I couldn't work out what it was conflicting with. I did have in mind to see if we can accept this one shift/reduce and understanding what the conflict is is a step in this direction. I also had in mind to see we could use some tricks I've recently noticed in the scanner, where the scanner does some simple forward parsing to change the tokens that it generates depending on the context. I find these parser problems usually a time sink with little reward though :(

I think method final

**Comment by ojwb** [comment]
> I think method final

Did you intend to finish that thought?

It occurs to me that SWIG supporting `final`as a keyword on classes but not as an identifier might be better than the current situation.  There's significant and increasing use of `final` on classes, but likely minimal and decreasing use of it as an identifier - although it's still valid, it feels unnatural now.

But a point against doing this is it might break a few existing SWIG bindings which work, and in a way I don't trivially see how to work around, so I think probably we need to accept the shift/reduce conflict, or find a way to avoid it.

More generally, while I understand that shift/reduce conflicts are undesirable, not parsing language features like `final` (over a decade old now) seems more undesirable.  I wonder if we need to come up with a clear methodology for how to decide when it is OK to introduce a new shift/reduce conflict, how to check how it's been resolved, how to adjust how it's resolved if need be, and how to document it.  Do we also need to add a testcase to ensure future parser changes don't change the resolution?  My understanding of bison is sadly rather limited.

**Comment by ojwb** [comment]
Hmm so we've actually had a user bug report for a bug where we didn't support `final` as an identifier, which makes my thought above about not supporting that less appealing.

**Comment by wsfulton** [comment]
I've been looking at this yesterday and today. I've experimented with various things to see if I can improve on it but have not so far. The patch as it is only has one problem that I have identified so far. I've made lots of testcases where 'final' could be used and so far the only problem is the conflict you identified, but in a very narrow set of cases::
```
struct Y final;
struct Y final = {};
struct Y final = Y();
```
where `struct` could also be `class` or `union`. Those are all variable declarations. Defining a variable like this is most unusual in C++ and only the first is valid C. The good news is the C case does not fail as we handle 'final' differently in C. So I think this is a small trade off to support final for class declarations. BTW, while I have tried many obscure test cases involving final, the obvious question is do these also fail in C++ (they don't!):
```
Y final;
int final;
enum A final;
```
Also, I have one last hacky idea to try out to fix it for C++, a simple state machine in the scanner for such a variable declaration to return ID instead of FINAL as the token. But I very much doubt it will be needed in the real world.

**Comment by ojwb** [comment]
That sounds great - `final` working in a class declaration is a worthwhile gain and losing `struct Y final;` in exchange seems an entirely reasonable trade off.

**Comment by devshgraphicsprogramming** [comment]
It still doesn't work for template classes with SWIG 4.0.2
```cpp
class Base
{
};


template<class T>
class Derived final : public Base
{
};
```

if you don't remove the `final` it will complain ` Error: Syntax error in input(1).`

@wsfulton will you reopen or should I open another issue?

**Comment by ojwb** [comment]
@devshgraphicsprogramming The fix wasn't in 4.0.2.  SWIG 4.1.0 parses your example.
